### ISSUE TITLE
[feature request] make supernode's CDN optional

### ISSUE THREAD

**Issue by xh4n3** [dragonflyoss/dragonfly-archived#356]
---
name: Feature Request - Make Supernode's CDN optional
about: CDN is good for protecting the content source. But in some scenarios, we may want to get rid of CDN. 

---

## Why you need it?
When we pull a large image among hundreds of nodes, the very first pieces are pulled from supernode. Only after peers claim they have pieces successfully downloaded, P2P network comes to work. 

If we have slow network on supernodes, (when it comes to hundreds of nodes, 100Mbps's bandwidth can be considered slow too), the supernode's bandwidth would be a bottleneck during the very first moment of pulling image in the cluster.

In our case, we host many supernodes for many different users in single VPC. When one user's many nodes are coming for the CDN's pieces, network usage will rise suddenly. This will somehow impact other users' network bandwidth. If we get rid of the CDN part, we can data flow and metadata flow, those sudden network spikes can be less often.

## How it could be?
Make supernode's CDN optional. 
When CDN is disabled, peers can pull from the source, then serve the content.

## Other related information
None

**Comment by allencloud** [comment]
Thanks for your feedback.

I am not sure if I understand your problem very well. Here you have put forward two parts of concern, right?

* supernode's bandwidth may be poor, then when it comes to be poor, it would be bottleneck under CDN functionality.
* one user's many nodes are coming for the CDN's pieces, network usage will rise suddenly.

Let us assume that I follow your problems. For first part, I think we can make supernode's bandwith more powerful. Is it possible? 
 
The second part, if we restrict the maximum client number of supernode, will it work for you? 

@xh4n3

**Comment by xh4n3** [comment]
Yes, IMHO, these two parts are potential disadvantages when CDN enabled, from reliability perspective.

Say we have 1Gbps's bandwidth on supernode, if we need to pull a 10G image via dragonfly. Peers need to wait about two minutes to begin download. (Please correct me if wrong)

Increasing supernode's bandwidth can be helpful, but there are still physical limits. When it comes to large clusters, hundreds of nodes are waiting for this single supernode's download.

Limiting dragonfly client number is a practical solution, but my customers wouldn't like it. 😭 

I have read the supernode's code, the CDN functionality is built through the whole lifecycle, it would be a big reconstruction to make CDN optional like a plugin.

@allencloud 

--

I implemented myself a version getting rid of the CDN. When dfget received pieces from supernode, dfget download files from source, and claim to supernode that they have downloaded the pieces, then serve the pieces to other peers. By this hacky way, dfget will not download from CDN.

On the other hand, I am considering limit bandwidth on the nginx in front of supernode.

**Comment by xq2248** [comment]
The newst dfget version can be called with "-p source", then you may download file directly from the source site. In this situation, dfget Client can not serve other peers.

**Comment by allencloud** [comment]
> The newst dfget version can be called with "-p source", then you may download file directly from the source site. In this situation, dfget Client can not serve other peers.

What you said is correct. Directly download the full file from the source, while no P2P services at all. @xq2248

**Comment by allencloud** [comment]
@xh4n3 

We totally support what you proposed in this issue. To make it happen, we have several plans. Maybe we can discuss on it to find which one is better.

1. support `dfget -p source+p2p` to make dfget downloads the file directly from the source address. and this dfget will still plays a role of peer to provide service for other peers. In addition, the supernode still contains CDN manager, while at this scenario the cdn manager will not work at all.

> dfget has a flag `-p`/`--pattern`, we add `source+p2p` in it.

| --pattern parameter | usage | 
| ------ | ------ |
|source | dfget directly downloads file from source address without p2p at all | 
|cdn|dfget directly downloads file from supernode's cdn without p2p at all|
|p2p|dfget downloads file supernode's cdn, and supernode schedules peers info to use p2p |
| source+p2p | dfget download pieces from source address and make supernode schedule them with p2p among peers network | 

2. enable a flag in supernode, like `supernode --enable-cdn=false`, then the supernode will start without cdn manager functionality. At this case, the dfget still sends request to supernode, and in the response from supernode, dfget will find that supernode has not enabled cdn functionality. Then it starts to download from the source address directly.

WDYT? Any input? @xh4n3

**Comment by xh4n3** [comment]
I think both of them are great! Plan 1 let the user determine how to use dragonfly while Plan 2 let the supernode's admin make the decision.

In my own case, I would like to use Plan 2, so my users wouldn't have the choice to corrupt my host cluster's network. For community, Plan 1 might be more popular?

So how about we merge two plans together? If supernode's flag is unset, i.e., true by default, let user determine how this download goes. When supernode's flag is set to false, only `source` pattern is allowed.

BTW, I think we can ignore some small files to go through p2p network cause the there is some time overhead.

**Comment by allencloud** [comment]
> So how about we merge two plans together?

Thanks for your kind suggestion. At first we have not realized that this a way. Now we agree it is much more convenient for user.

We will take you advice. @xh4n3 @Starnop @lowzj

**Comment by Starnop** [comment]
> So how about we merge two plans together? If supernode's flag is unset, i.e., true by default, let user determine how this download goes. When supernode's flag is set to false, only `source` pattern is allowed.

@xh4n3  FYI, #480. 

If there are any suggestions, please feel free to put forward it. THX.
